### ISSUE TITLE
Bugsnag breaks app with CSP in place

### ISSUE THREAD

**Issue by dpolivy** [bugsnag/bugsnag-js#357]
### Expected behavior
I expect Bugsnag to load just fine with a CSP that follows the [documentation](https://docs.bugsnag.com/platforms/browsers/faq/#can-i-use-bugsnag-with-csp).

### Observed behavior
I get errors that `sessions.bugsnag.com` is not part of my CSP and the connection is refused. Even worse, on iOS 9/10, this completely breaks the loading of my app!

### Steps to reproduce
Setup Bugsnag with a CSP as directed in the FAQ. Then, try and load it.

### Version
v4 from the CDN link. Not sure which specific version it is.

### Additional information
This seemed to break last week sometime.

This is a *huge* issue, and I'm disappointed that more notice wasn't given for a change like this.

**Comment by bengourley** [comment]
Hi @dpolivy

Thanks for raising this. I have been able to reproduce the issue in an older version of iOS (9 – but I could not reproduce in the minor version of 10 that I tried: 10.3) where the CSP violations prevent the rest of a script from loading.

CSP violations are not supposed to cause exceptions, however it turns out that in these versions of Safari they do. At some point in iOS 10 the behaviour was fixed – it now behaves in a spec-compliant way and per other browsers by just logging an error (rather than throwing a `DOMException`).

We've identified a fix for the issue and will release it in due course. As part of that fix we'll be adding some CSP test cases to our end-to-end suite to prevent regressions in this area. As you also spotted, the docs were not updated and they too will be rectified.

As far as pro-actively giving notice around releases goes, we weren't aware of this browser inconsistency – if we were we would have shipped the workaround along with the changes. For any know "breaking" changes we would bump the major version (currently 4).

As you say, you're using the v4 CDN link which just grabs the latest minor/patch version as soon as it is released – I would recommend pinning this to a specific minor/patch version and to manually opt-in to upgrades. For example:

**Pinned to minor version 4.7, only accepting patch updates**
https://d2wy8f7a9ursnm.cloudfront.net/v4.7/bugsnag.min.js

**Pinned to exact version 4.7.1**
https://d2wy8f7a9ursnm.cloudfront.net/v4.7.1/bugsnag.min.js

[More detail here](https://docs.bugsnag.com/platforms/browsers/js/versioning/)

That way you can review [the changelog](https://github.com/bugsnag/bugsnag-js/blob/master/CHANGELOG.md) for each release and decide what (if any) the risks of upgrading are, or whether you would need to do some additional testing. We strive to do the best we can in terms of QA to ensure things like this don't happen, but we can't simulate every kind of environment, so ultimately while our aim is to minimise the likelihood of a release having issues, it is always a possibility.

Thanks again for bringing this to our attention and massive apologies for the inconvenience it has caused you.

tl;dr
- we've reproduced the problem
- we've identified a fix
- we'll ship the fix and update the outdated docs
- we recommend pinning the version
- sorry, and thanks!

**Comment by dpolivy** [comment]
@bengourley Awesome, thanks for following up and addressing this. I'm happy to hear you're adding test cases around this to prevent any future issues, and adjusting the code to handle it more gracefully. Ironically, Bugsnag didn't seem to catch these errors so it took a bit of manual sleuthing to track down the problem 😄 

I appreciate your detailed response and the follow-up actions to improve from this. Thank you!
